File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/howto/samples/xml/xmlschemaobjectmodel/vb/XmlSchemaObjectModel.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft .NET SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'-----------------------------------------------------------------------

Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Xml
Imports System.Xml.Schema

namespace HowTo.Samples.XML

public class XmlSchemaObjectModelSample

    Dim myXmlSchema as XmlSchema 
    Dim myXmlTextWriter as XmlTextWriter 

    'Schemas to read
    private const sampleSchema1 as String = "books.xsd"
    private const sampleSchema2 as String = "poschema.xsd"

    public shared sub Main()
        Dim args As String() = {sampleSchema1, sampleSchema2}
        Dim MyXmlSchemaObjectModelSample as XmlSchemaObjectModelSample
        MyXmlSchemaObjectModelSample = new XmlSchemaObjectModelSample ()
        MyXmlSchemaObjectModelSample.Run(args)
    end sub

    public sub Run(args As String())
        try
            Console.WriteLine ("Creating Schema in the Schema Object Model...")
 
            myXmlTextWriter = new XmlTextWriter(Console.Out)
            myXmlTextWriter.Formatting = Formatting.Indented
            myXmlTextWriter.Indentation = 2
            'Create an XmlNameTable
            Dim myXmlNameTable as XmlNameTable  = new NameTable()
            'Add the nametable to the XmlSchemaCollection
            Dim myXmlSchemaCollection as XmlSchemaCollection  = new XmlSchemaCollection(myXmlNameTable)
            'Add some schemas to the XmlSchemaCollection
            Console.WriteLine ("Reading and adding {0} schema.", args(0))
            myXmlSchemaCollection.Add(nothing, args(0))
            Console.WriteLine ("Reading and adding {0} schema.", args(1))
            myXmlSchemaCollection.Add(nothing, args(1))
            Console.WriteLine ("Added schemas successfully ...")

            Console.WriteLine ("Showing added schemas")
            
            Dim myTempXmlSchema as XmlSchema
            for each myTempXmlSchema in myXmlSchemaCollection 
                myXmlSchema = myTempXmlSchema
                Console.WriteLine()
                Console.WriteLine("Schema from: {0}", myTempXmlSchema.SourceUri)
                Console.WriteLine()
                Console.WriteLine("=== Start of Schema ===")
                Console.WriteLine()

                ' Write out the various schema parts
                WriteXSDSchema()

                Console.WriteLine()
                Console.WriteLine()
                Console.WriteLine("=== End of Schema ===")
                Console.WriteLine()
                Console.WriteLine("Example of possible XML contents for: {0}", myTempXmlSchema.SourceUri)
                Console.WriteLine()
                Console.WriteLine("=== Start of Example ===")

                ' Write out an example of the XML for the schema
                WriteExample()

                Console.WriteLine()
                Console.WriteLine()
                Console.WriteLine("=== End of Example ===")
            next

        catch e as Exception
            Console.WriteLine ("Exception: {0}", e.ToString())
        end try
    end sub

    ' Write out the XSD 
    private sub WriteXSDSchema() 
        myXmlTextWriter.WriteStartElement("schema", XmlSchema.Namespace)
        myXmlTextWriter.WriteAttributeString("targetNamespace", myXmlSchema.TargetNamespace)
        Dim include as XmlSchemaInclude
        for each include in myXmlSchema.Includes 
            myXmlTextWriter.WriteStartElement("include", XmlSchema.Namespace)
            myXmlTextWriter.WriteAttributeString("schemaLocation", include.SchemaLocation)
            myXmlTextWriter.WriteEndElement()
        next
        
        Dim item as object
        for each item in myXmlSchema.Items 

            if (TypeOf item is XmlSchemaAttribute) 
                WriteXmlSchemaAttribute(CType(item, XmlSchemaAttribute)) 'attribute
            else if (TypeOf item is XmlSchemaComplexType) 
                WriteXmlSchemaComplexType(CType(item, XmlSchemaComplexType)) 'complexType
            else if (TypeOf item is XmlSchemaSimpleType) 
                WriteXmlSchemaSimpleType(CType(item, XmlSchemaSimpleType)) 'simpleType
            else if (TypeOf item is XmlSchemaElement) 
                WriteXmlSchemaElement(CType(item, XmlSchemaElement)) 'element
            else if (TypeOf item is XmlSchemaAnnotation) 
                WriteXmlSchemaAnnotation(CType(item, XmlSchemaAnnotation)) 'annotation
            else if (TypeOf item is XmlSchemaAttributeGroup) 
                WriteXmlSchemaAttributeGroup(CType(item, XmlSchemaAttributeGroup)) 'attributeGroup
            else if (TypeOf item is XmlSchemaNotation) 
                WriteXmlSchemaNotation(CType(item, XmlSchemaNotation)) 'notation
            else if (TypeOf item is XmlSchemaGroup) 
                WriteXmlSchemaGroup(CType(item, XmlSchemaGroup)) 'group
            else 
                Console.WriteLine("Not Implemented.")
            end if
        next
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaAttribute
    private sub WriteXmlSchemaAttribute(attribute as XmlSchemaAttribute ) 

        myXmlTextWriter.WriteStartElement("attribute", XmlSchema.Namespace)
        if not attribute.Name is nothing 
            myXmlTextWriter.WriteAttributeString("name", attribute.Name)
        end if
        if (not attribute.RefName.IsEmpty) 
            myXmlTextWriter.WriteStartAttribute("ref", nothing)
            myXmlTextWriter.WriteQualifiedName(attribute.RefName.Name, attribute.RefName.Namespace)
            myXmlTextWriter.WriteEndAttribute()
        end if
        if (not attribute.SchemaTypeName.IsEmpty) 
            myXmlTextWriter.WriteStartAttribute("type", nothing)
            myXmlTextWriter.WriteQualifiedName(attribute.SchemaTypeName.Name, attribute.SchemaTypeName.Namespace)
            myXmlTextWriter.WriteEndAttribute()
        end if
        if not attribute.SchemaType is nothing 
            WriteXmlSchemaSimpleType(attribute.SchemaType)
        end if
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaComplexType
    private sub WriteXmlSchemaComplexType(complexType as XmlSchemaComplexType ) 

        myXmlTextWriter.WriteStartElement("complexType", XmlSchema.Namespace)
        if not complexType.Name is nothing 
            myXmlTextWriter.WriteAttributeString("name", complexType.Name)
        end if
        if not complexType.ContentModel is nothing 
            Console.WriteLine("Not Implemented for this ContentModel.")
        else 
            if not complexType.Particle is nothing 
                WriteXmlSchemaParticle(complexType.Particle)
            end if
            Dim o as object
            for each o in complexType.Attributes
                if (TypeOf o is XmlSchemaAttribute) 
                    WriteXmlSchemaAttribute(CType(o, XmlSchemaAttribute))
                end if
            next
        end if
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaSimpleType
    private sub WriteXmlSchemaSimpleType(simpleType as XmlSchemaSimpleType ) 

        myXmlTextWriter.WriteStartElement("simpleType", XmlSchema.Namespace)
        if not simpleType.Name is nothing
            myXmlTextWriter.WriteAttributeString("name", simpleType.Name)
        end if
        if (TypeOf simpleType.Content is XmlSchemaSimpleTypeRestriction) 
            myXmlTextWriter.WriteStartElement("restriction", XmlSchema.Namespace)
        else if (TypeOf simpleType.Content is XmlSchemaSimpleTypeList) 
            myXmlTextWriter.WriteStartElement("list", XmlSchema.Namespace)
        else 
            myXmlTextWriter.WriteStartElement("union", XmlSchema.Namespace)
        end if
        myXmlTextWriter.WriteEndElement()
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaParticle
    private sub WriteXmlSchemaParticle(particle as XmlSchemaParticle ) 
        if (TypeOf particle is XmlSchemaElement) 
            WriteXmlSchemaElement(CType(particle, XmlSchemaElement))
        else if (TypeOf particle is XmlSchemaSequence) 
            myXmlTextWriter.WriteStartElement("sequence", XmlSchema.Namespace)
            Dim particle1 as XmlSchemaParticle
            for each particle1 in CType(particle, XmlSchemaSequence).Items
                WriteXmlSchemaParticle(particle1)
            next
            myXmlTextWriter.WriteEndElement()
        else 
            Console.WriteLine("Not Implemented for this type: {0}", particle.ToString())
        end if
    end sub

    ' XmlSchemaElement
    private sub WriteXmlSchemaElement(element as XmlSchemaElement ) 
        myXmlTextWriter.WriteStartElement("element", XmlSchema.Namespace)
        if not element.Name is nothing 
            myXmlTextWriter.WriteAttributeString("name", element.Name)
        end if
        if (not element.RefName.IsEmpty) 
            myXmlTextWriter.WriteStartAttribute("ref", nothing)
            myXmlTextWriter.WriteQualifiedName(element.RefName.Name, element.RefName.Namespace)
            myXmlTextWriter.WriteEndAttribute()
        end if
        if (not element.SchemaTypeName.IsEmpty) 
            myXmlTextWriter.WriteStartAttribute("type", nothing)
            myXmlTextWriter.WriteQualifiedName(element.SchemaTypeName.Name, element.SchemaTypeName.Namespace)
            myXmlTextWriter.WriteEndAttribute()
        end if
        if not element.SchemaType is nothing
            if (TypeOf element.SchemaType is XmlSchemaComplexType) 
                WriteXmlSchemaComplexType(CType(element.SchemaType, XmlSchemaComplexType))
            else 
                WriteXmlSchemaSimpleType(CType(element.SchemaType, XmlSchemaSimpleType))
            end if
        end if
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaAnnotation 
    private sub WriteXmlSchemaAnnotation(annotation as XmlSchemaAnnotation ) 
        ' Not a complete implementation
        myXmlTextWriter.WriteStartElement("annotation", XmlSchema.Namespace)
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaAttributeGroup
    private sub WriteXmlSchemaAttributeGroup(attributeGroup as XmlSchemaAttributeGroup ) 
        ' Not a complete implementation
        myXmlTextWriter.WriteStartElement("attributeGroup", XmlSchema.Namespace)
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaGroup
    private sub WriteXmlSchemaGroup(group as XmlSchemaGroup ) 
        ' Not a complete implementation
        myXmlTextWriter.WriteStartElement("group", XmlSchema.Namespace)
        myXmlTextWriter.WriteEndElement()
    end sub

    'XmlSchemaNotation
    private sub WriteXmlSchemaNotation(notation as XmlSchemaNotation ) 
        ' Not a complete implementation
        myXmlTextWriter.WriteStartElement("notation", XmlSchema.Namespace)
        myXmlTextWriter.WriteEndElement()
    end sub

    ' Write out the example of the XSD usage
    private sub WriteExample() 
        Dim element as XmlSchemaElement
        for each element in myXmlSchema.Elements.Values
            WriteExampleElement(element)
        next
    end sub
    
    ' Write some example elements
    private sub WriteExampleElement(element as XmlSchemaElement ) 
        myXmlTextWriter.WriteStartElement(element.QualifiedName.Name, element.QualifiedName.Namespace)
        if (TypeOf element.ElementType is XmlSchemaComplexType) 
            Dim type as XmlSchemaComplexType  = CType(element.ElementType, XmlSchemaComplexType)
            if not type.ContentModel is nothing 
                Console.WriteLine("Not Implemented for this ContentModel")
            end if
            WriteExampleAttributes(type.Attributes)
            WriteExampleParticle(type.Particle)
        else 
            WriteExampleValue(element.ElementType)
        end if
        myXmlTextWriter.WriteEndElement()
    end sub

    ' Write some example attributes
    private sub WriteExampleAttributes(attributes as XmlSchemaObjectCollection )
        Dim o as object 
        for each o in attributes
            if (TypeOf o is XmlSchemaAttribute) 
                 WriteExampleAttribute(CType(o, XmlSchemaAttribute))
            else 
                Dim group as XmlSchemaAttributeGroup
                Dim oTemp as XmlSchemaAttributeGroupRef 
                oTemp = CType(o, XmlSchemaAttributeGroupRef)
                group = CType(myXmlSchema.Groups(oTemp.RefName), XmlSchemaAttributeGroup)
                WriteExampleAttributes(group.Attributes)
            end if
        next
    end sub

    ' Write a single example attribute
    private sub WriteExampleAttribute(attribute as XmlSchemaAttribute ) 
        myXmlTextWriter.WriteStartAttribute(attribute.QualifiedName.Name, attribute.QualifiedName.Namespace)
        ' The examples value
        WriteExampleValue(attribute.AttributeType)
        myXmlTextWriter.WriteEndAttribute()
    end sub

    ' Write example particles
    private sub WriteExampleParticle(particle as XmlSchemaParticle ) 

        Dim max as long  
        if (particle.MaxOccurs = -1 OR particle.MaxOccurs > 10000)
            max = 5
        else
            max = particle.MaxOccurs
        end if

        Dim i as integer
        for i = 0 to CInt(max) - 1 
            if (TypeOf particle is XmlSchemaElement) 
                WriteExampleElement(CType(particle, XmlSchemaElement))
            else if (TypeOf particle is XmlSchemaSequence) 
                Dim particle1 as XmlSchemaParticle
                for each particle1 in CType(particle, XmlSchemaSequence).Items
                    WriteExampleParticle(particle1)
                next
            else 
                Console.WriteLine("Not Implemented for this type: {0}", particle.ToString())
            end if
        next i
    end sub

    ' Write the examples text values
    private sub WriteExampleValue(schemaType as object ) 
        Dim datatype as XmlSchemaDatatype 
        if (TypeOf schemaType is XmlSchemaSimpleType)
            datatype = (CType(schemaType, XmlSchemaSimpleType)).Datatype 
        else
            datatype = CType(schemaType, XmlSchemaDatatype)
        end if

        ' Consult the XSD to CLR conversion table for the correct type mappings
        Dim type as Type  = datatype.ValueType

        if (type is GetType(Boolean)) 
            myXmlTextWriter.WriteString("true")
        else if ((type is GetType(integer)) or (type is GetType(long))) 
            myXmlTextWriter.WriteString("12345")
        else if ((type is GetType(single)) or (type is GetType(decimal))) 
            myXmlTextWriter.WriteString("10.99")
        else if (type is GetType(System.Xml.XmlQualifiedName)) 
            myXmlTextWriter.WriteString("qualified_name")
        else if (type is GetType(System.DateTime)) 
            myXmlTextWriter.WriteString("12-12-2001")
        else if (type is GetType(string)) 
            myXmlTextWriter.WriteString("ExampleString")
        ' Handle the 'xsd:positiveInteger' XSD type in the SOMsample.xsd
        else if (type is GetType(System.UInt64)) 
            'positiveInteger
            myXmlTextWriter.WriteString("10")
        else 
            myXmlTextWriter.WriteString("Not Implemented for this datatype: " + datatype.ToString())
            myXmlTextWriter.WriteString(type.FullName)

        end if
    end sub

end Class 'End class XmlSchemaObjectModel
end Namespace 'End namespace HowTo.Samples.XML

File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/howto/samples/xml/xmlschemaobjectmodel/cs/XmlSchemaObjectModel.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft .NET SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//-----------------------------------------------------------------------

namespace HowTo.Samples.XML
{

using System;
using System.IO;
using System.Xml;
using System.Xml.Schema;

public class XmlSchemaObjectModelSample
{
    Random random = new Random();
    XmlSchema myXmlSchema;
    XmlTextWriter myXmlTextWriter;

    //Schemas to read
    private const String sampleSchema1 = "books.xsd";
    private const String sampleSchema2 = "poschema.xsd";

    public static void Main()
    {
        String[] args = {sampleSchema1, sampleSchema2};
        XmlSchemaObjectModelSample MyXmlSchemaObjectModelSample = new XmlSchemaObjectModelSample();
        MyXmlSchemaObjectModelSample.Run(args);
    }

    public void Run(String[] args)
    {
        try
        {
            Console.WriteLine ("Creating Schema in the Schema Object Model...");

            myXmlTextWriter = new XmlTextWriter(Console.Out);
            myXmlTextWriter.Formatting = Formatting.Indented;
            myXmlTextWriter.Indentation = 2;
            //Create an XmlNameTable
            XmlNameTable myXmlNameTable = new NameTable();
            //Add the nametable to the XmlSchemaCollection
            XmlSchemaCollection myXmlSchemaCollection = new XmlSchemaCollection(myXmlNameTable);
            //Add some schemas to the XmlSchemaCollection
            Console.WriteLine ("Reading and adding {0} schema.", args[0]);
            myXmlSchemaCollection.Add(null, args[0]);
            Console.WriteLine ("Reading and adding {0} schema.", args[1]);
            myXmlSchemaCollection.Add(null, args[1]);
            Console.WriteLine ("Added schemas successfully ...");

            Console.WriteLine ("Showing added schemas");

            foreach(XmlSchema myTempXmlSchema in myXmlSchemaCollection)
            {
                myXmlSchema = myTempXmlSchema;
                Console.WriteLine();
                Console.WriteLine("Schema from: {0}", myTempXmlSchema.SourceUri);
                Console.WriteLine();
                Console.WriteLine("=== Start of Schema ===");
                Console.WriteLine();

                // Write out the various schema parts
                WriteXSDSchema();

                Console.WriteLine();
                Console.WriteLine();
                Console.WriteLine("=== End of Schema ===");
                Console.WriteLine();
                Console.WriteLine("Example of possible XML contents for: {0}", myTempXmlSchema.SourceUri);
                Console.WriteLine();
                Console.WriteLine("=== Start of Example ===");

                // Write out an example of the XML for the schema
                WriteExample();

                Console.WriteLine();
                Console.WriteLine();
                Console.WriteLine("=== End of Example ===");
            }

        }
        catch (Exception e)
        {
            Console.WriteLine ("Exception: {0}", e.ToString());
        }
    }

    // Write out the XSD
    void WriteXSDSchema()
    {
        myXmlTextWriter.WriteStartElement("schema", XmlSchema.Namespace);
        myXmlTextWriter.WriteAttributeString("targetNamespace", myXmlSchema.TargetNamespace);
        foreach(XmlSchemaInclude include in myXmlSchema.Includes)
        {
            myXmlTextWriter.WriteStartElement("include", XmlSchema.Namespace);
            myXmlTextWriter.WriteAttributeString("schemaLocation", include.SchemaLocation);
            myXmlTextWriter.WriteEndElement();
        }

        foreach(object item in myXmlSchema.Items)
        {
            if (item is XmlSchemaAttribute)
                WriteXmlSchemaAttribute((XmlSchemaAttribute)item);      //attribute
            else if (item is XmlSchemaComplexType)
                WriteXmlSchemaComplexType((XmlSchemaComplexType)item);  //complexType
            else if (item is XmlSchemaSimpleType)
                WriteXmlSchemaSimpleType((XmlSchemaSimpleType)item);    //simpleType
            else if (item is XmlSchemaElement)
                WriteXmlSchemaElement((XmlSchemaElement)item);          //element
            else if (item is XmlSchemaAnnotation)
                WriteXmlSchemaAnnotation((XmlSchemaAnnotation)item);    //annotation
            else if (item is XmlSchemaAttributeGroup)
                WriteXmlSchemaAttributeGroup((XmlSchemaAttributeGroup)item); //attributeGroup
            else if (item is XmlSchemaNotation)
                WriteXmlSchemaNotation((XmlSchemaNotation)item);        //notation
            else if (item is XmlSchemaGroup)
                WriteXmlSchemaGroup((XmlSchemaGroup)item);              //group
            else
                Console.WriteLine("Not Implemented.");
        }
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaAttribute
    void WriteXmlSchemaAttribute(XmlSchemaAttribute attribute)
    {
        myXmlTextWriter.WriteStartElement("attribute", XmlSchema.Namespace);
        if (attribute.Name != null)
            myXmlTextWriter.WriteAttributeString("name", attribute.Name);

        if (!attribute.RefName.IsEmpty)
        {
            myXmlTextWriter.WriteStartAttribute("ref", null);
            myXmlTextWriter.WriteQualifiedName(attribute.RefName.Name, attribute.RefName.Namespace);
            myXmlTextWriter.WriteEndAttribute();
        }

        if (!attribute.SchemaTypeName.IsEmpty)
        {
            myXmlTextWriter.WriteStartAttribute("type", null);
            myXmlTextWriter.WriteQualifiedName(attribute.SchemaTypeName.Name, attribute.SchemaTypeName.Namespace);
            myXmlTextWriter.WriteEndAttribute();
        }

        if (attribute.SchemaType != null)
            WriteXmlSchemaSimpleType(attribute.SchemaType);

        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaComplexType
    void WriteXmlSchemaComplexType(XmlSchemaComplexType complexType)
    {
        myXmlTextWriter.WriteStartElement("complexType", XmlSchema.Namespace);
        if (complexType.Name != null)
            myXmlTextWriter.WriteAttributeString("name", complexType.Name);

        if (complexType.ContentModel != null)
            Console.WriteLine("Not Implemented for this ContentModel.");
        else
        {
            if (complexType.Particle != null)
                WriteXmlSchemaParticle(complexType.Particle);
            foreach(object o in complexType.Attributes)
            {
                if (o is XmlSchemaAttribute)
                    WriteXmlSchemaAttribute((XmlSchemaAttribute)o);
            }
        }
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaSimpleType
    void WriteXmlSchemaSimpleType(XmlSchemaSimpleType simpleType)
    {
        myXmlTextWriter.WriteStartElement("simpleType", XmlSchema.Namespace);
        if (simpleType.Name != null)
            myXmlTextWriter.WriteAttributeString("name", simpleType.Name);

        if (simpleType.Content is XmlSchemaSimpleTypeRestriction)
            myXmlTextWriter.WriteStartElement("restriction", XmlSchema.Namespace);
        else if (simpleType.Content is XmlSchemaSimpleTypeList)
            myXmlTextWriter.WriteStartElement("list", XmlSchema.Namespace);
        else
            myXmlTextWriter.WriteStartElement("union", XmlSchema.Namespace);

        myXmlTextWriter.WriteEndElement();
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaParticle
    void WriteXmlSchemaParticle(XmlSchemaParticle particle)
    {
        if (particle is XmlSchemaElement)
            WriteXmlSchemaElement((XmlSchemaElement)particle);
        else if (particle is XmlSchemaSequence)
        {
            myXmlTextWriter.WriteStartElement("sequence", XmlSchema.Namespace);
            foreach(XmlSchemaParticle particle1 in ((XmlSchemaSequence)particle).Items)
                WriteXmlSchemaParticle(particle1);

            myXmlTextWriter.WriteEndElement();
        }
        else
            Console.WriteLine("Not Implemented for this type: {0}", particle.ToString());
    }

    // XmlSchemaElement
    void WriteXmlSchemaElement(XmlSchemaElement element)
    {
        myXmlTextWriter.WriteStartElement("element", XmlSchema.Namespace);
        if (element.Name != null)
            myXmlTextWriter.WriteAttributeString("name", element.Name);

        if (!element.RefName.IsEmpty)
        {
            myXmlTextWriter.WriteStartAttribute("ref", null);
            myXmlTextWriter.WriteQualifiedName(element.RefName.Name, element.RefName.Namespace);
            myXmlTextWriter.WriteEndAttribute();
        }

        if (!element.SchemaTypeName.IsEmpty)
        {
            myXmlTextWriter.WriteStartAttribute("type", null);
            myXmlTextWriter.WriteQualifiedName(element.SchemaTypeName.Name, element.SchemaTypeName.Namespace);
            myXmlTextWriter.WriteEndAttribute();
        }

        if (element.SchemaType != null)
        {
            if (element.SchemaType is XmlSchemaComplexType)
                WriteXmlSchemaComplexType((XmlSchemaComplexType)element.SchemaType);
            else
                WriteXmlSchemaSimpleType((XmlSchemaSimpleType)element.SchemaType);
        }
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaAnnotation
    void WriteXmlSchemaAnnotation(XmlSchemaAnnotation annotation)
    {
        // Not a complete implementation
        myXmlTextWriter.WriteStartElement("annotation", XmlSchema.Namespace);
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaAttributeGroup
    void WriteXmlSchemaAttributeGroup(XmlSchemaAttributeGroup attributeGroup)
    {
        // Not a complete implementation
        myXmlTextWriter.WriteStartElement("attributeGroup", XmlSchema.Namespace);
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaGroup
    void WriteXmlSchemaGroup(XmlSchemaGroup group)
    {
        // Not a complete implementation
        myXmlTextWriter.WriteStartElement("group", XmlSchema.Namespace);
        myXmlTextWriter.WriteEndElement();
    }

    //XmlSchemaNotation
    void WriteXmlSchemaNotation(XmlSchemaNotation notation)
    {
        // Not a complete implementation
        myXmlTextWriter.WriteStartElement("notation", XmlSchema.Namespace);
        myXmlTextWriter.WriteEndElement();
    }

    // Write out the example of the XSD usage
    void WriteExample()
    {
        foreach(XmlSchemaElement element in myXmlSchema.Elements.Values)
            WriteExampleElement(element);
    }

    // Write some example elements
    void WriteExampleElement(XmlSchemaElement element)
    {
        myXmlTextWriter.WriteStartElement(element.QualifiedName.Name, element.QualifiedName.Namespace);
        if (element.ElementType is XmlSchemaComplexType)
        {
            XmlSchemaComplexType type = (XmlSchemaComplexType)element.ElementType;
            if (type.ContentModel != null)
                Console.WriteLine("Not Implemented for this ContentModel");
            WriteExampleAttributes(type.Attributes);
            WriteExampleParticle(type.Particle);
        }
        else
            WriteExampleValue(element.ElementType);

        myXmlTextWriter.WriteEndElement();
    }

    // Write some example attributes
    void WriteExampleAttributes(XmlSchemaObjectCollection attributes)
    {
        foreach(object o in attributes)
        {
            if (o is XmlSchemaAttribute)
                 WriteExampleAttribute((XmlSchemaAttribute)o);
            else
            {
                XmlSchemaAttributeGroup group = (XmlSchemaAttributeGroup)myXmlSchema.Groups[((XmlSchemaAttributeGroupRef)o).RefName];
                WriteExampleAttributes(group.Attributes);
            }
        }
    }

    // Write a single example attribute
    void WriteExampleAttribute(XmlSchemaAttribute attribute)
    {
        myXmlTextWriter.WriteStartAttribute(attribute.QualifiedName.Name, attribute.QualifiedName.Namespace);
        // The examples value
        WriteExampleValue(attribute.AttributeType);
        myXmlTextWriter.WriteEndAttribute();
    }

    // Write example particles
    void WriteExampleParticle(XmlSchemaParticle particle)
    {
        Decimal max;

        if (particle.MaxOccurs == -1 || particle.MaxOccurs > 10000)
            max = 5;
        else
            max = particle.MaxOccurs;

        for (int i = 0; i < max; i ++)
        {
            if (particle is XmlSchemaElement)
                WriteExampleElement((XmlSchemaElement)particle);
            else if (particle is XmlSchemaSequence)
            {
                foreach(XmlSchemaParticle particle1 in ((XmlSchemaSequence)particle).Items)
                    WriteExampleParticle(particle1);
            }
            else
                Console.WriteLine("Not Implemented for this type: {0}", particle.ToString());
        }
    }

    // Write the examples text values
    void WriteExampleValue(object schemaType)
    {
        XmlSchemaDatatype datatype = (schemaType is XmlSchemaSimpleType) ? ((XmlSchemaSimpleType)schemaType).Datatype : (XmlSchemaDatatype)schemaType;

        // Consult the XSD to CLR conversion table for the correct type mappings
        Type type = datatype.ValueType;
        if (type == typeof(bool))
            myXmlTextWriter.WriteString("true");
        else if (type == typeof(int) || type == typeof(long))
            myXmlTextWriter.WriteString(random.Next(100).ToString());
        else if (type == typeof(float) || type == typeof(decimal))
            myXmlTextWriter.WriteString(((float)random.Next(100)/(float)random.Next(10, 20)).ToString("#.00"));
        else if (type == typeof(System.Xml.XmlQualifiedName))
            myXmlTextWriter.WriteString("qualified_name" + random.Next(100).ToString());
        else if (type == typeof(DateTime))
            myXmlTextWriter.WriteString("12-12-2001");
        else if (type == typeof(string))
            myXmlTextWriter.WriteString("ExampleString" + random.Next(100).ToString());
        // Handle the 'xsd:positiveInteger' XSD type in the SOMsample.xsd
        else if (type == typeof(System.UInt64))
            //positiveInteger
            myXmlTextWriter.WriteString(random.Next(100).ToString());
        else
            myXmlTextWriter.WriteString("Not Implemented for this datatype: " + datatype.ToString());
    }

} // End class XmlSchemaObjectModel
} // End namespace HowTo.Samples.XML